### PR TITLE
multi: fix chain notif. blocking by pmt processor.

### PR DESCRIPTION
**Pull request by dnldd**
This fixes chain notification messages being blocked by payment processing calls when tx confirmations are not accurately being reported.

The fix includes:

- ensuring a payment processing call concludes before starting another one via the processing flag.

- calling the payment process in a goroutine to avoid blocking the chain state process.

- keeping track of tx confirmation failures and starting a wallet rescan if the failures exceed threshold.

Associated tests have been updated.

### CODE REVIEW

File: pool/paymentmgr.go [hunk 1]
```diff
@@ -148,6 +166,21 @@ func NewPaymentMgr(pCfg *PaymentMgrConfig) (*PaymentMgr, error) {
...
+func (pm *PaymentMgr) isProcessing() bool {
+	pm.mtx.Lock()
+	defer pm.mtx.Unlock()
+	return pm.processing
+}
+
+// setProcessing sets the processing flag to the provided boolean.
+func (pm *PaymentMgr) setProcessing(status bool) {
+	pm.mtx.Lock()
+	pm.processing = status
```
Comment (jholdstock) [inline pool/paymentmgr.go:180]: Using a Mutex to protect a bool like this seems unnecessary. The mutex itself can be used to wait until any on-going processing has finished.

File: pool/paymentmgr.go [hunk 2]
```diff
@@ -644,13 +730,51 @@ func (pm *PaymentMgr) payDividends(ctx context.Context, height uint32, treasuryA
 		return nil
 	}
 
+	if pm.isProcessing() {
+		log.Info("Waiting for in-progress payment processing to " +
```
Comment (jholdstock) [inline pool/paymentmgr.go:734]: This isnt going to wait if `nil` is returned here. It will terminate the payment processing and stop it proceeding. Simply waiting for the mutex to be unlocked would fix this issue.

File: pool/paymentmgr.go [hunk 3]
```diff
@@ -644,13 +730,51 @@ func (pm *PaymentMgr) payDividends(ctx context.Context, height uint32, treasuryA
...
+			"complete before starting another")
+		return nil
+	}
+
+	pm.setProcessing(true)
```
Comment (jholdstock) [inline pool/paymentmgr.go:739]: Just do `.Lock()` here and defer `.Unlock()`

File: pool/paymentmgr.go (hunk 1 above)
Comment (dnldd) [reply pool/paymentmgr.go:180]: The plan is to have any subsequent calls while the flag is set terminate immediately, not wait to run in the future.

File: pool/paymentmgr.go (hunk 2 above)
Comment (dnldd) [reply pool/paymentmgr.go:734]: I can see why you thought I wanted the call to wait and execute, will update the comment.

File: pool/paymentmgr.go (hunk 3 above)
Comment (dnldd) [reply pool/paymentmgr.go:739]: not looking to hold the lock and have subsequent calls wait to execute.

File: pool/chainstate.go [hunk 4]
```diff
@@ -398,6 +386,18 @@ func (cs *ChainState) handleChainUpdates(ctx context.Context) {
 					cs.cfg.Cancel()
 					continue
 				}
+
+				// Process mature payments without blocking.
+				go func(cfg *ChainStateConfig, height uint32, treasuryActive bool) {
```
Comment (jholdstock) [inline pool/chainstate.go:391]: Why is this moved down? I don't think this is going to run unless the most recently mined block is one mined by the pool? (As we discussed previously on Matrix)

Comment (dnldd) [reply pool/chainstate.go:391]: Good catch, noticed the current call wasn't only running for pool mining mode, that's why it was moved. Rectifying.

**Comment by jholdstock** [comment]
Tried deploying the latest, along with my additional logging, and I'm seeing this:

```
2021-05-04 13:49:27.865 [TRC] POOL: Requesting tx conf notifications for 18 transactions (stop after 676684)
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL: Got notification with 0 confirmations
2021-05-04 13:49:27.865 [TRC] POOL: 18 coinbases are not confirmed:
2021-05-04 13:49:27.865 [TRC] POOL:     128cb9805fb29af374a755a938eb33848f9f4fe44c108cc3a47eeb1eb598a105
2021-05-04 13:49:27.865 [TRC] POOL:     b3a417c505e9fec327511397390b42e249083ec44ada12896e46e1439dce30c6
2021-05-04 13:49:27.865 [TRC] POOL:     044baa51a549d4a6d350e0ee6fab683f204950c271199eca0a37f38c0894cb9d
2021-05-04 13:49:27.865 [TRC] POOL:     a0ba191745a9b46429c4a9e84d5b521847d60797f1809ba9288a543f4424a504
2021-05-04 13:49:27.865 [TRC] POOL:     48628bf8a39363286360cad41b62ee426780e99b43dfea101f3945f37d25ea35
2021-05-04 13:49:27.865 [TRC] POOL:     e5be2fcf3a6e2015e0e755212ccc744f658585b836eb6bade48836c8c766c671
2021-05-04 13:49:27.865 [TRC] POOL:     c5f9419f567c38220fdfbaf245b0223292a9233a377dffd7215a76e9d0f067d2
2021-05-04 13:49:27.865 [TRC] POOL:     9b6228cce3311c1a00bf95f4d78d59de5e09ec5aa54a002d5d2d84ff7e14095b
2021-05-04 13:49:27.865 [TRC] POOL:     1952eb7463a3b2d147d5abdf47e190a97e12f6bffd3f6c849795fec7fd51458c
2021-05-04 13:49:27.865 [TRC] POOL:     08df22d7ada1e3d18cda26df5ed2de5045f01a05632a6496f803648e32985687
2021-05-04 13:49:27.865 [TRC] POOL:     a5ecb55c4891fa3ddbc39c5cd9a4d49caec1966ad05829a620408cab09fe416f
2021-05-04 13:49:27.865 [TRC] POOL:     b55022510b40befe7ec551ea35cb22a8effa594cad327769c034550cb3a2450e
2021-05-04 13:49:27.865 [TRC] POOL:     227845cec3ed12c2bd15512c48777805b300a71507240a9e5f85fbf94bdae3d8
2021-05-04 13:49:27.865 [TRC] POOL:     cb65f4151e52fb20d4e7c9a1bc64910c5d9ff862ec5325cf730b3c9440e2cb1d
2021-05-04 13:49:27.865 [TRC] POOL:     6d382962125ed768e501b077d0b8be75d0f4e63717ea4b81a82544366b80a4be
2021-05-04 13:49:27.865 [TRC] POOL:     6a61fbf760f5165147227e89afe4a2fb3b3c992a4f1a7278f53b43f3baf46650
2021-05-04 13:49:27.865 [TRC] POOL:     7962b13efb1f579191cfd8a4a83d72b65c74cf1dffc219959b980ae65b4a95d9
2021-05-04 13:49:27.865 [TRC] POOL:     5c2def82127965621ef2f05d6ad39effabd6635628873fc5472945b6473d6920
```

File: pool/hub.go [hunk 5]
```diff
@@ -442,10 +445,13 @@ func (h *Hub) getWork(ctx context.Context) (string, string, error) {
 
 // getTxConfNotifications streams transaction confirmation notifications for
 // the provided transaction hashes.
-func (h *Hub) getTxConfNotifications(txHashes []*chainhash.Hash, stopAfter int32) (func() (*walletrpc.ConfirmationNotificationsResponse, error), error) {
+func (h *Hub) getTxConfNotifications(txHashes []chainhash.Hash, stopAfter int32) (func() (*walletrpc.ConfirmationNotificationsResponse, error), error) {
 	hashes := make([][]byte, 0, len(txHashes))
+	log.Tracef("Requesting tx conf notifications for %d "+
+		"transactions (stop after #%d)", len(txHashes), stopAfter)
```
Comment (jholdstock) [inline pool/hub.go:451]: Checking docs for `stopAfter` at https://github.com/decred/dcrwallet/blob/master/rpc/documentation/api.md#confirmationnotifications

> `int32 stop_after`: The upper limit on how many confirmations is needed before the transaction becomes unwatched.

We are passing `spendableHeight` as `stopAfter` which means notifications will continue until each tx has a huge number of confs. We only need to wait for 1 conf I think?

File: pool/paymentmgr_test.go [hunk 6]
```diff
@@ -1244,3 +1410,183 @@ func testPaymentMgrDust(t *testing.T) {
...
+			createRawTransaction: func(ctx context.Context, inputs []chainjson.TransactionInput, amounts map[dcrutil.Address]dcrutil.Amount, lockTime *int64, expiry *int64) (*wire.MsgTx, error) {
+				return &wire.MsgTx{}, nil
+			},
+		}
+	}
+	mgr.cfg.GetBlockConfirmations = func(ctx context.Context, bh *chainhash.Hash) (int64, error) {
+		return int64(estMaturity) + 1, nil
+	}
+
+	// Ensure the payment lifecycle process recieves the payment signal and
```
Comment (jholdstock) [inline pool/paymentmgr_test.go:1550]: receives*

**Review by jholdstock — CHANGES_REQUESTED** [review]
Completed a thorough review, comments attached.

As we have discussed on Matrix, I am really not a fan of this change because we are adding a large amount of code to work around a bug in dcrwallet, which should really be fixed there. In order to ensure we come back and remove this code later, I would like to propose two things:

- We add a good, lengthy comment to the code, to explain the dcrwallet bug and why the rescan is necessary.
- We immediately open a new issue on the dcrpool repo to remove this code as soon as decred/dcrwallet#1740 is fixed. It should link to this PR and to the dcrwallet issue for reference.

File: pool/paymentmgr.go [hunk 7]
```diff
@@ -788,5 +983,44 @@ func (pm *PaymentMgr) payDividends(ctx context.Context, height uint32, treasuryA
...
+	pm.mtx.Lock()
+	pm.txConfHashes = make(map[chainhash.Hash]uint32)
+	pm.mtx.Unlock()
+
 	return nil
 }
+
+// processPayments relays payment signals for processing.
+func (pm *PaymentMgr) processPayments(msg *paymentMsg) {
+	pm.paymentCh <- msg
```
[the remark this answers is not in the document]
Comment (dnldd) [reply pool/paymentmgr.go:998]: not sure what you're referring to here.

File: pool/paymentmgr_test.go [hunk 8]
```diff
@@ -169,12 +182,21 @@ func createPaymentMgr(paymentMethod string) (*PaymentMgr, error) {
...
 		PoolFeeAddrs:          []dcrutil.Address{poolFeeAddrs},
+		Cancel:                cancel,
+		SignalCache:           signalCache,
+		HubWg:                 new(sync.WaitGroup),
 	}
-	return NewPaymentMgr(pCfg)
+
+	mgr, err := NewPaymentMgr(pCfg)
+	if err != nil {
+		return nil, nil, nil, err
```
[the remark this answers is not in the document]
Comment (dnldd) [reply pool/paymentmgr_test.go:192]: `createPaymentMgr` isn't taking `t` as a parameter, it's a helper.

File: pool/paymentmgr_test.go [hunk 9]
```diff
@@ -1128,6 +1157,129 @@ func testPaymentMgrPayment(t *testing.T) {
...
+		}
+	}
+
+	err = mgr.payDividends(ctx, estMaturity+1, treasuryActive)
+	if !errors.Is(err, errs.Rescan) {
+		cancel()
+		t.Fatalf("expected a rescan error, got %v", err)
+	}
+
+	// Ensure dividend payment returns an error if fetching rescan responses fail.
```
[the remark this answers is not in the document]
Comment (dnldd) [reply pool/paymentmgr_test.go:1211]: good catch.

File: pool/chainstate.go [hunk 10]
```diff
@@ -31,8 +31,8 @@ type ChainStateConfig struct {
 	db Database
 	// SoloPool represents the solo pool mining mode.
 	SoloPool bool
-	// PayDividends pays mature mining rewards to participating accounts.
-	PayDividends func(context.Context, uint32, bool) error
+	// ProcessPayments relays payment signals for Processing.
```
[the remark this answers is not in the document]
Comment (jholdstock) [reply pool/chainstate.go:34]: This comment is resolved but I still see the typo

File: pool/paymentmgr.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jholdstock) [reply pool/paymentmgr.go:998]: Misplaced comment, resolving.

File: pool/paymentmgr_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jholdstock) [reply pool/paymentmgr_test.go:192]: but we could pass it in? I think its worthwhile, we will remove 6 duplicated instances of `createPaymentMgr` error handling

[the remark this answers is not in the document]
Comment (jholdstock) [reply pool/paymentmgr_test.go:192]: But we could pass it in? I think it would be worthwhile, it would remove 6 duplicated instances of `createPaymentMgr` error handling

**Review by jholdstock — CHANGES_REQUESTED** [review]

File: pool/chainstate.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (dnldd) [reply pool/chainstate.go:34]: recitifying.

File: pool/paymentmgr_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (dnldd) [reply pool/paymentmgr_test.go:192]: valid, updating.

**Review by jholdstock — APPROVED** [review]
